flink-ml-core/src/main/java/org/apache/flink/ml/common/datastream/DataStreamUtils.java

```diff
@@ -67,6 +72,28 @@ public static <IN, OUT> DataStream<OUT> mapPartition(
                 .setParallelism(input.getParallelism());
     }
 
+    /**
+     * Applies a {@link ReduceFunction} on a bounded data stream. The output stream contains at most
+     * one stream record and its parallelism is one.
+     *
+     * @param input The input data stream.
+     * @param func The user defined reduce function.
+     * @param <T> The class type of the input.
+     * @return The result data stream.
+     */
+    public static <T> DataStream<T> reduce(DataStream<T> input, ReduceFunction<T> func) {
+        DataStream<T> partialReducedStream =
+                input.transform("partialReduce", input.getType(), new ReduceOperator<>(func))
+                        .setParallelism(input.getParallelism());
```

Comment:
It might be better to use `forward()` and `colocate()` here so that the reduce is of better performance.

Reply:
Shouldn't these two operators be chained together?

Reply:
So far as I can see from the `JobGraph` of LinearRegressionTest.testFitAndPredict`, `reduce` and `partialReduce` belongs to different `JobVertex` and not chained together yet.

Reply:
Yeah, you are right and it is not chained. But from the job graph, I could see that the edge is already `FORWARD` and there is no network shuffle.